src/bg/objects/user-script-downloader.js

```diff
@@ -0,0 +1,182 @@
+/* Download content from the provided URLS */
+
+class ScriptDownloader {
+  constructor(contentInfo, progressId=null) {
+    this.contentInfo = contentInfo;
+    this.progressId = progressId;
+
+    this._downloads = new Map();
+    this._scriptDetails = {'requiresContent': {}, 'resources': {}};
+  }
+
+  run() {
+    let downloads =
+        this.contentInfo.map(cInfo => {
+          let download = this._buildDownload(cInfo);
+          return download.run()
+              .then(this._wrapFunction(this._handleResult, cInfo));
+        });
+
+    return Promise.all(downloads).then(() => {
+      // One last progress report
+      if (this.progressId) {
+        let progress = Array(this._downloads.size).fill(1);
+        ScriptInstall.reportDownloadProgress(this.progressId, progress);
+      }
+      return this._scriptDetails;
+    });
+  }
+
+  abort() {
+    this._downloads.forEach(download => download.abort());
+  }
+
+  sendProgress() {
+    if (this.progressId) {
+      let progress = [];
+      this._downloads.forEach(download => progress.push(download.progress));
+      ScriptInstall.reportDownloadProgress(this.progressId, progress);
+    }
+  }
+
+  _buildDownload(cInfo) {
+    let download;
+    switch (cInfo.type) {
+      case 'require':
+        // Fetch as a string
+        download = new Download(cInfo.url, false, this);
+        break;
+      case 'icon':
+      case 'resource':
+        // Fetch as a blob
+        download = new Download(cInfo.url, true, this);
+    }
+    this._downloads.set(cInfo, download);
+    return download;
+  }
+
+  // Format script details based on the items that were downloaded
+  _handleResult(result, cInfo) {
+    switch (cInfo.type) {
+      case 'icon':
+        this._scriptDetails.iconBlob = result;
+        break;
+
+      case 'require':
+        this._scriptDetails.requiresContent[cInfo.url] = result;
+        break;
+
+      case 'resource':
+        this._scriptDetails.resources[cInfo.key] = {
+          'name': cInfo.key,
+          'mimetype': this._downloads.get(cInfo).contentType,
+          'url': cInfo.url,
+          'blob': result
+        };
+        break;
+    }
+  }
+
+  _wrapFunction(fn, ...args) {
+    return data => {
+      if (data !== undefined) {
+        args.unshift(data);
+      }
+      fn.apply(this, args);
```

Comment:
Shoudn't a real wrapper `return` that value here? `apply` itself does...

Reply:
Probably. I didn't think about it because all the wrapped functions don't return anything. Honestly, I probably don't need it in the `ScriptDownloader` class as it's only used in one place.

Reply:
I guessed so! No need for this in `onWhatever` callbacks, but unnecessary surprising, if used for other purposes...